### ISSUE TITLE
Webhook Payload Types are Wrong (in Patreon's official documentation)

### ISSUE THREAD

**Issue by boylett** [ghostrider-05/patreon-api.ts#89]
### Bug description

1. Create a new webhook in Patreon
2. Set it to post to a [https://webhook.site/](https://webhook.site/) endpoint
3. Post a test trigger (`members:pledge:create` or `members:pledge:update`)
4. Observe structure that differs from official documentation (and thus `patreon-api.ts` types)

### Expected behaviour

Patreon's documentation is notoriously terrible - so I don't blame you at all for the typings being a little off - so please observe the attached payload example. I hope this helps!

### System settings

Node version: 22.8.0
Npm version: 10.8.2
Patreon-api.ts version: 0.7.0

### Potential fixes

Please observe the following payload as an example of a `members:pledge:create` or `members:pledge:update` trigger payload. Both triggers provide the same structure and information.

Some notes:
- Patreon's documentation claims that user data comes through as `data.relationships.patron` - but this is false. User data actually comes through as `data.relationships.user`.
- `last_charge_status` is always either `"Paid"` or `"Pending"`.
- Hvaing received several thousand payloads, I have never seen `data.attributes.patron_status` have a value other than `"active_patron"`.
- Pretty much every `data.relationships.attributes.social_connections` entry I've seen has been an object containing just `user_id` (string) and `url` (string).
- I've never seen `data.attributes.pledge_cadence` with anything other than `1`.

```json
{
  "data": {
    "attributes": {
      "campaign_lifetime_support_cents": 0,
      "currently_entitled_amount_cents": 0,
      "email": "someone@example.com",
      "full_name": "User Fullname",
      "is_follower": false,
      "is_free_trial": false,
      "last_charge_date": "2024-09-02T19:55:30.000+00:00",
      "last_charge_status": "Pending",
      "lifetime_support_cents": 0,
      "next_charge_date": "2024-10-02T00:00:00.000+00:00",
      "note": "",
      "patron_status": "active_patron",
      "pledge_cadence": 1,
      "pledge_relationship_start": "2024-09-02T19:55:28.617+00:00",
      "will_pay_amount_cents": 450
    },
    "id": "efefefef-efef-efef-efef-efefefefefef",
    "relationships": {
      "address": {
        "data": null
      },
      "campaign": {
        "data": {
          "id": "12312312",
          "type": "campaign"
        },
        "links": {
          "related": "https://www.patreon.com/api/oauth2/v2/campaigns/12312312"
        }
      },
      "currently_entitled_tiers": {
        "data": [
          {
            "id": "12312312",
            "type": "tier"
          }
        ]
      },
      "user": {
        "data": {
          "id": "123123123",
          "type": "user"
        },
        "links": {
          "related": "https://www.patreon.com/api/oauth2/v2/user/123123123"
        }
      }
    },
    "type": "member"
  },
  "included": [
    {
      "attributes": {
        "created_at": "2023-10-26T12:13:56.000+00:00",
        "creation_name": "",
        "discord_server_id": "1231231231231231231",
        "google_analytics_id": null,
        "has_rss": false,
        "has_sent_rss_notify": false,
        "image_small_url": "https://...",
        "image_url": "https://...",
        "is_charged_immediately": true,
        "is_monthly": true,
        "is_nsfw": false,
        "main_video_embed": null,
        "main_video_url": null,
        "one_liner": null,
        "patron_count": 219,
        "pay_per_name": "month",
        "pledge_url": "/checkout/vanity-slug",
        "published_at": "2023-10-26T12:16:12.000+00:00",
        "rss_artwork_url": null,
        "rss_feed_title": null,
        "summary": "<p style=\"\">...</p>",
        "thanks_embed": null,
        "thanks_msg": "<p style=\"\">...</p>",
        "thanks_video_url": null,
        "url": "https://www.patreon.com/vanity-slug",
        "vanity": "vanity-slug"
      },
      "id": "12312312",
      "type": "campaign"
    },
    {
      "attributes": {
        "about": "",
        "created": "2024-09-02T19:45:18.000+00:00",
        "first_name": "User",
        "full_name": "Firstname",
        "hide_pledges": false,
        "image_url": "https://...",
        "is_creator": false,
        "last_name": "",
        "like_count": 0,
        "social_connections": {
          "discord": {
            "scopes": [
              "identify",
              "email"
            ],
            "user_id": "",
            "url": ""
          },
          "facebook": {
            "user_id": "",
            "url": ""
          },
          "google": {
            "user_id": "",
            "url": ""
          },
          "instagram": {
            "user_id": "",
            "url": ""
          },
          "reddit": {
            "user_id": "",
            "url": ""
          },
          "spotify": {
            "user_id": "",
            "url": ""
          },
          "spotify_open_access": {
            "user_id": "",
            "url": ""
          },
          "tiktok": {
            "user_id": "",
            "url": ""
          },
          "twitch": {
            "user_id": "",
            "url": ""
          },
          "twitter": {
            "user_id": "",
            "url": ""
          },
          "twitter2": {
            "user_id": "",
            "url": ""
          },
          "vimeo": {
            "user_id": "",
            "url": ""
          },
          "youtube": {
            "user_id": "",
            "url": ""
          }
        },
        "thumb_url": "https://...",
        "url": "https://...",
        "vanity": null
      },
      "id": "123123123",
      "type": "user"
    }
  ],
  "links": {
    "self": "https://www.patreon.com/api/oauth2/v2/members/efefefef-efef-efef-efef-efefefefefef"
  },
}
```

### Additional context

I run and maintain a modestly sized community site that utilizes Patreon as a means of generating revenue. If you need any further examples of webhooks or data structures please feel free to ask, I'd love to see this library flourish and take away the pain involved when working with Patreon's disasterous API!

**Comment by ghostrider-05** [comment]
Hi, first of all thank you for believing in this project and the ⭐. I came to the conclusion that the `member:pledge` resource type should be `member` and not `pledge-event`, that will be fixed in `v0.7.1`. I can test some of the API, but I don't have patrons or use this library in campaigns with a lot of members so I really appreciate the help! If you want to use this for your site, I'm open for feature requests / direct messages if there's something you want in this library.

While implementing the webhooks, I had mainly a doubt about the included attributes: can I assume that all attributes (with the creator token) will be sent?
Some reactions to your notes:
- Does this happen only with (pledge) webhooks or with all pledge event resources? I'm not a fan of implementing undocumented resource relationships, but it's weird that it is different. If the resource type of pledge webhooks are indeed `member` (seems confirmed by your example), then the docs are correct.
- Will add a note in the docs, but leaving all documented values in there for now
- Can't confirm that with testing on a campaign of <10 members. Seeing all values from null to declined_patron
- Will improve the types for `social_connections` in `v0.8.0`, this example payload will help a lot!
- Will add a note in the docs

**Comment by boylett** [comment]
Will happily oblige there. I've got big plans for some future Patreon integrations, so as soon as I get to those I'm sure I'll have some notes for you!

Not sure what you mean regarding the creator token; the fields you see in the example above are everything I've been receiving from `members:pledge:create` and `members:pledge:update` webhooks.

Regarding your points;
- It is weird, and I hate it too. But I have been receiving all of these fields regularly for the last year or so with no deviation that I can find.
- 👍 
- Interesting! Now we know some of the possible field types :'D

I'll keep an eye out as my work continues. Thanks for creating this lib, it's working like a dream so far!

**Comment by ghostrider-05** [comment]
If you are using the creator tokens in the dev portal from Patreon you will include all scopes and thus access to all fields, from the documentation:
> The client creator’s access token will automatically have all V2 scopes associated with it.

I made the assumption that all attributes will be sent and typed it as `CompletePayload<PatreonWebhookTrigger>`, which seems to be correct looking at the payload you sent.

- If I change the type of `member:pledge:x` webhooks to `WebhookPayload<Patreon>`, I think all your problems are solved as it fixes:
   - `data.type` is correctly typed as `member`
   - user relationship is under `data.relationships.user`. The Patreon documentation is still correct for the Pledge resource with the `patron` relationship
- These are the documented field types, so I'm leaving `patron_status` as it is.

**Comment by ghostrider-05** [comment]
I've create a PR that should fix everything you mentioned:

- `member:pledge:x` webhooks will be correctly typed
- some fields have a note for more explanation
- `User.social_connections` can be set to any type through module augmentation, defaults to `object`

**Comment by ghostrider-05** [comment]
Hey, this is now released in [v0.8.0](https://github.com/ghostrider-05/patreon-api.ts/releases/tag/patreon-api.ts-v0.8.0)! Sorry for the (long) wait since this was implemented.

**Comment by boylett** [comment]
Awesome stuff!! :D
